// filestack-js/build/module/index.js
/*
 * Copyright (c) 2018 by Filestack.
 * Some rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { registerModule, FILESTACK_MODULES } from '@filestack/loader';
import { getSecurity, validateWebhookSignature } from './lib/api/security';
import { Client } from './lib/client';
import { PickerDisplayMode } from './lib/picker';
import { TransformSchema } from './schema/transforms.schema';
import { PrefetchEvents } from './lib/api/prefetch';
import { FilestackError } from './filestack_error';
import { getMimetype } from './lib/utils/index';
/**
 * Initialize client with given config
 *
 * @param apikey
 * @param options
 */
export var Filestack = function (apikey, options) {
    return new Client(apikey, options);
};
// This will be deprecated in feature use
export var init = Filestack;
/**
 * filestack-js version. Interpolated at build time.
 */
export var version = '3.16.0';
registerModule(FILESTACK_MODULES.FILESTACK_SDK, Filestack, { version: '3.16.0' });
export * from './lib/api/transform';
export * from './lib/filelink';
export * from './filestack_error';
export * from './lib/request';
export * from './lib/utils';
export { TransformSchema, getSecurity, validateWebhookSignature, Client, FilestackError, PickerDisplayMode, getMimetype, PrefetchEvents };
