// filestack-js/build/main/lib/request/token.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// token
var token = function (data) { return new Promise(function (resolve) { return data.listeners.push(resolve); }); };
var tokenSource = function () {
    var data = {
        reason: null,
        listeners: [],
    };
    var cancel = function (reason) {
        reason = reason || 'Aborted';
        if (typeof reason === 'string') {
            reason = new Error(reason);
        }
        data.reason = reason;
        // Only for security reason
        /* istanbul ignore next */
        setTimeout(function () {
            for (var i = 0; i < data.listeners.length; i++) {
                if (typeof data.listeners[i] === 'function') {
                    data.listeners[i](reason);
                    data.listeners.splice(i, 1);
                }
            }
        }, 0);
    };
    return {
        cancel: cancel,
        token: token(data),
    };
};
/**
 * Filestack token that allow pause, resume or cancel given upload
 *
 * @export
 * @class FsToken
 * @extends {EventEmitter}
 * @implements {FsTokenInterface}
 */
var FsCancelToken = /** @class */ (function () {
    function FsCancelToken() {
        var cancelable = tokenSource();
        this.source = cancelable.token;
        this.cancelMethod = cancelable.cancel;
    }
    /**
     * Cancel request action
     *
     * @param {(string | Error)} [reason]
     * @memberof Token
     */
    FsCancelToken.prototype.cancel = function (reason) {
        this.cancelMethod(reason);
    };
    /**
     * Returns cancel token promise
     *
     * @returns
     * @memberof Token
     */
    FsCancelToken.prototype.getSource = function () {
        return this.source;
    };
    return FsCancelToken;
}());
exports.FsCancelToken = FsCancelToken;
