// filestack-js/build/main/lib/request/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * Copyright (c) 2018 by Filestack
 * Some rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var FsHttpMethod;
(function (FsHttpMethod) {
    FsHttpMethod["GET"] = "GET";
    FsHttpMethod["DELETE"] = "DELETE";
    FsHttpMethod["HEAD"] = "HEAD";
    FsHttpMethod["PURGE"] = "PURGE";
    FsHttpMethod["OPTIONS"] = "OPTIONS";
    FsHttpMethod["POST"] = "POST";
    FsHttpMethod["PUT"] = "PUT";
    FsHttpMethod["PATH"] = "PATH";
})(FsHttpMethod = exports.FsHttpMethod || (exports.FsHttpMethod = {}));
